Mathlib/Tactic/CategoryTheory/Monoidal.lean

```diff
@@ -0,0 +1,681 @@
+    let f := η.src
+    let f' := η.tar
+    let ⟨η₃, pf_η₃⟩ ← evalComp η₂ (.associator f' g h)
+    let ⟨η₄, pf_η₄⟩ ← evalComp (.associatorInv f g h) η₃
+    try return ⟨η₄, ← mkAppM ``evalWhiskerRight_comp #[pf_η₁, pf_η₂, pf_η₃, pf_η₄]⟩
+    catch _ => return ⟨η₄, mkConst ``True⟩
+  | η, .id => do
+    let f := η.src
+    let g := η.tar
+    let ⟨η₁, pf_η₁⟩ ← evalComp η (.rightUnitorInv g)
+    let ⟨η₂, pf_η₂⟩ ← evalComp (.rightUnitor f) η₁
+    try return ⟨η₂, ← mkAppM ``evalWhiskerRight_id #[pf_η₁, pf_η₂]⟩
+    catch _ => return ⟨η₂, mkConst ``True⟩
+
+/-- Evaluate the expression of a 2-morphism into a normalized form. -/
+partial def eval (e : Expr) : MonoidalM Result := do
+  if let .some α ← structuralAtom? e then
+    try return ⟨.nil <| .atom α, ← mkEqRefl (← α.e)⟩
+    catch _ => return ⟨.nil <| .atom α, mkConst ``True⟩
+  else
+    match e.getAppFnArgs with
+    | (``CategoryStruct.id, #[_, _, f]) =>
+      try return ⟨.nil (.id (← toMor₁ f)), ← mkEqRefl (← mkAppM ``CategoryStruct.id #[f])⟩
+      catch _ => return ⟨.nil (.id (← toMor₁ f)), mkConst ``True⟩
+    | (``CategoryStruct.comp, #[_, _, _, _, _, η, θ]) =>
+      let ⟨η_e, pf_η⟩ ← eval η
+      let ⟨θ_e, pf_θ⟩ ← eval θ
+      let ⟨ηθ, pf⟩ ← evalComp η_e θ_e
+      try return ⟨ηθ, ← mkAppM ``eval_comp #[pf_η, pf_θ, pf]⟩
+      catch _ => return ⟨ηθ, mkConst ``True⟩
+    | (``MonoidalCategoryStruct.whiskerLeft, #[_, _, _, f, _, _, η]) =>
+      let ⟨η_e, pf_η⟩ ← eval η
+      let ⟨θ, pf_θ⟩ ← evalWhiskerLeftExpr (← toMor₁ f) η_e
+      try return ⟨θ, ← mkAppM ``eval_whiskerLeft #[pf_η, pf_θ]⟩
+      catch _ => return ⟨θ, mkConst ``True⟩
+    | (``MonoidalCategoryStruct.whiskerRight, #[_, _, _, _, _, η, h]) =>
+      let ⟨η_e, pf_η⟩ ← eval η
+      let ⟨θ, pf_θ⟩ ← evalWhiskerRightExpr η_e (← toMor₁ h)
+      try return ⟨θ, ← mkAppM ``eval_whiskerRight #[pf_η, pf_θ]⟩
+      catch _ => return ⟨θ, mkConst ``True⟩
+    | (``monoidalComp, #[C, _, _, _, _, _, _, η, θ]) =>
+      let ⟨η_e, pf_η⟩ ← eval η
+      let α₀ ← structuralOfMonoidalComp C e
+      let α := NormalExpr.nil α₀
+      let ⟨θ_e, pf_θ⟩ ← eval θ
+      let ⟨αθ, pf_θα⟩ ← evalComp α θ_e
+      let ⟨ηαθ, pf_ηαθ⟩ ← evalComp η_e αθ
+      try return ⟨ηαθ, ← mkAppM ``eval_monoidalComp #[pf_η, pf_θ, pf_θα, pf_ηαθ]⟩
+      catch _ => return ⟨ηαθ, mkConst ``True⟩
+    | _ =>
+      try return ⟨← NormalExpr.ofExpr e, ← mkAppM ``eval_of #[e]⟩
+      catch _ => return ⟨← NormalExpr.ofExpr e, mkConst ``True⟩
+
+/-- Convert a `NormalExpr` expression into a list of `WhiskerLeftExpr` expressions. -/
+def NormalExpr.toList : NormalExpr → List WhiskerLeftExpr
+  | NormalExpr.nil _ => []
+  | NormalExpr.cons _ η ηs => η :: NormalExpr.toList ηs
+
+end Mathlib.Tactic.Monoidal
+
+open Mathlib.Tactic.Monoidal
+
+/-- `normalize% η` is the normalization of the 2-morphism `η`. It is of the form
+`α₀ ≫ η₀ ≫ α₁ ≫ η₁ ≫ ... αₙ ≫ ηₙ ≫ αₙ₊₁`, where `αᵢ` are structural 2-morphisms
+and `ηᵢ` are non-structural 2-morphisms. -/
```

Comment:
This description of normal forms seems a bit less specific than the description in the module doc at the top. Does the more specific description apply here?